doc/plugins.rst

```diff
@@ -0,0 +1,303 @@
+method has a entry_points keywords with a group ``mapproxy`` pointing to
+a module with a ``plugin_entrypoint`` method.
+
+.. code-block:: python
+
+    entry_points={"mapproxy": ["hips = mapproxy_hips.pluginmodule"]},
+
+
+In this example, the mapproxy_hips/pluginmodule.py file should have
+a ``plugin_entrypoint`` method taking no argument and returning nothing.
+
+.. code-block:: python
+
+    def plugin_entrypoint():
+        # call different registration methods, like register_service_configuration(),
+        # register_source_configuration()
+        pass
+
+That method is in charge of registering the various registration methods
+detailed hereafter.
+
+Plugins will often by dependent on MapProxy internal classes. It is their
+responsibility to check the MapProxy version, in case the MapProxy internal
+API or behavior would change and make them incompatible.
+
+Adding a new service
+--------------------
+
+The ``mapproxy.config.loader`` module has a ``register_service_configuration()``
+method to register a new service and specify the allowed keywords for it in
+the YAML configuration file.
+
+.. code-block:: python
+
+    def register_service_configuration(service_name, service_creator,
+                                       yaml_spec_service_name = None, yaml_spec_service_def = None):
+        """ Method used by plugins to register a new service.
+
+            :param config_name: Name of the service
+            :type config_name: str
+            :param service_creator: Creator method of the service
+            :type service_creator: method of type (serviceConfiguration: ServiceConfiguration, conf: dict) -> Server
+            :param yaml_spec_service_name: Name of the service in the YAML configuration file
+            :type yaml_spec_service_name: str
+            :param yaml_spec_service_def: Definition of the service in the YAML configuration file
+            :type yaml_spec_service_def: dict
+        """
+
+
+This can for example by used like the following snippet:
+
+.. code-block:: python
+
+    from mapproxy.config.loader import register_service_configuration
+    from mapproxy.service.base import Server
+
+    class MyExtraServiceServer(Server):
+        # Look at classes at https://github.com/mapproxy/mapproxy/tree/master/mapproxy/service
+        # for a real implementation
+        names = ('my_extra_service',)
+        def __init__(self):
+            pass
+
+    def my_extra_service_method(serviceConfiguration, conf):
+        return MyExtraServiceServer()
+
+    register_service_configuration('my_extra_service', my_extra_service_method,
+                                   'my_extra_service', {'foo': str()})
+
+
+This allows the following declaration in the YAML mapproxy configuration file:
+
+.. code-block:: yaml
+
+    services:
+        my_extra_service:
+            foo: bar
+
+A real-world implementation can be found at https://github.com/rouault/mapproxy_hips/blob/master/mapproxy_hips/service/hips.py
```

Comment:
If I understand the tests below correctly, this also includes tests with minimnal example plugin. This is awesome.